library/platform_util.c

```diff
@@ -65,3 +72,45 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
     memset_func( buf, 0, len );
 }
 #endif /* MBEDTLS_PLATFORM_ZEROIZE_ALT */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE) && !defined(MBEDTLS_PLATFORM_GMTIME_ALT)
+#include <time.h>
+#if !defined(_WIN32) && (defined(__unix__) || \
+    (defined(__APPLE__) && defined(__MACH__)))
+#include <unistd.h>
+#if !defined(_POSIX_VERSION) || _POSIX_C_SOURCE > _POSIX_THREAD_SAFE_FUNCTIONS
+#define PLATFORM_UTIL_USE_GMTIME
+#endif /* !_POSIX_VERSION || _POSIX_C_SOURCE > _POSIX_THREAD_SAFE_FUNCTIONS */
+#endif /* !_WIN32 && (__unix__ || (__APPLE__ && __MACH__)) */
+
+struct tm *mbedtls_platform_gmtime( const mbedtls_time_t *tt,
+                                    struct tm *tm_buf )
+{
+#if defined(_WIN32) && !defined(EFIX64) && !defined(EFI32)
+    return( ( gmtime_s( tm_buf, tt ) == 0 ) ? tm_buf : NULL );
+#elif !defined(PLATFORM_UTIL_USE_GMTIME)
+    return( gmtime_r( tt, tm_buf ) );
+#else
+    struct tm *lt;
+
+#if defined(MBEDTLS_THREADING_C)
```

Comment:
Is this check for `defined(MBEDTLS_THREADING_C)` necessary? The reason `gmtime()` was originally removed by @NWilson is because it's not thread safe, and if we're putting `gmtime()` back in, it needs to be threadsafe, and having locks shouldn't be optional.

Better in my view, to remove them and add to `check_config.h` a dependency on `MBEDTLS_THREADING_C` if `MBEDTLS_HAVE_TIME_DATE` is defined, for platforms that don't have `gmtime()`.

Reply:
While in principle I agree with this view. Forcing MBEDTLS_HAVE_TIME_DATE to depend on MBEDTLS_THREADING_C changes the API. Also, consider that not every platform that uses MBEDTLS_HAVE_TIME_DATE will be multithreaded.

Reply:
I think the code is fine as it is: as Andres said, `gmtime()` is perfectly fine for single-threaded platforms, and for multi-threaded platforms, there are already plenty of reasons why they need to define `MBEDTLS_THREADING_C` anyway, so this change is not introducing any new requirement. We never promised to work in threaded environments without `MBEDTLS_THREADING_C`, quite the contrary.

Reply:
I've changed my mind on this, because we do run on some single threaded systems that don't have `gmtime_r()`, but `gmtime()` is available. I think it's fine as it is.